pkg/api/imagebasedgroupupgrades/v1alpha1/types.go

```diff
@@ -0,0 +1,104 @@
+
+func (i ImageBasedUpgradeAction) String() string {
+	return i.Name
+}
+
+var (
+	// Prep defines the preparing stage for image based upgrade
+	Prep = ImageBasedUpgradeAction{"Prep"}
+	// Upgrade defines the upgrading stage for image based upgrade
+	Upgrade = ImageBasedUpgradeAction{"Upgrade"}
+	// Rollback defines the rollback stage for image based upgrade
+	Rollback = ImageBasedUpgradeAction{"Rollback"}
+	// Finalize defines the finalizing stage for image based upgrade
+	Finalize = ImageBasedUpgradeAction{"Finalize"}
+	// Abort defines the aborting stage for image based upgrade
+	Abort = ImageBasedUpgradeAction{"Abort"}
+	// Unknown defines an unknown ImageBasedUpgrade action
+	Unknown = ImageBasedUpgradeAction{""}
+)
+
+// RolloutStrategy defines how to rollout ibu
+type RolloutStrategy struct {
+	//kubebuilder:validation:Minimum=1
+	MaxConcurrency int `json:"maxConcurrency"`
+	//+kubebuilder:default=240
+	Timeout int `json:"timeout,omitempty"`
+}
+
+// ImageBasedGroupUpgradeSpec defines the desired state of ImageBasedGroupUpgrade
+type ImageBasedGroupUpgradeSpec struct {
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="IBU Spec"
+	//+kubebuilder:validation:Required
+	IBUSpec lcav1alpha1.ImageBasedUpgradeSpec `json:"ibuSpec"`
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Clusters",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:text"}
+	Clusters []string `json:"clusters,omitempty"`
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Cluster Label Selectors",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:text"}
+	ClusterLabelSelectors []metav1.LabelSelector `json:"clusterLabelSelectors,omitempty"`
+	//kubebuilder:validation:Minimum=1
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Actions"
+	// kubebuilder:validation:XValidation:message="Invalid list of actions", rule="self.map(x,x.name)==['Prep'] || self.map(x,x.name)==['Prep','Upgrade'] || self.map(x,x.name)==['Prep','Upgrade','Finalize'] || self.map(x,x.name)==['Rollback'] || self.map(x,x.name)==['Rollback', 'Finalize'] || self.map(x,x.name)==['Upgrade'] || self.map(x,x.name)==['Upgrade','Finalize'] || self.map(x,x.name)==['Finalize'] || self.map(x,x.name)==['Abort']"
+	// +kubebuilder:validation:XValidation:message="You can only add actions to the list", rule="oldSelf.all(x, (x in self))"
+	// +kubebuilder:validation:MaxItems=4
+	Actions []ImageBasedUpgradeAction `json:"actions"`
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Rollout Strategy",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:text"}
+	//+kubebuilder:validation:Required
+	RolloutStrategy RolloutStrategy `json:"rolloutStrategy,omitempty"`
+}
+
+// ClusterState defines the current state of a cluster
+type ClusterState struct {
+	Name          string  `json:"name"`
+	State         string  `json:"state"`
+	CurrentAction *string `json:"CurrentAction,omitempty"`
+	Message       *string `json:"message,omitempty"`
```

Comment:
Should we bundle the three fields into a currentAction struct? In case we will add cluster level stuff later?

Reply:
Maybe a better way is to have a list of action status, matching the spec:
```
- name: cluster1
  - action: prep
    state: succeeded
  - action: upgrade
    state: in progress
- name: cluster2
  - action: prep
    state: failed
  - action: upgrade
    state: skipped or not selected
- name: cluster3
  - action: prep
    state: succeeded
  - action: upgrade
    state: succeeded
  - action: finalize
    state: in progress
```

Reply:
Probably too verbose though